R/shoji-catalog.R

```diff
@@ -240,28 +240,109 @@ setMethod("emails", "ShojiCatalog", function (x) getIndexSlot(x, "email"))
+        return(data.frame())
+    } else {
+        ### The following code is equivalent to out <- purrr::map_df(index, entry_to_df)
+        ################
+        entry_list <- lapply(index, entry_to_df, list_col_names = list_columns)
+        names   <- unique(unlist( lapply(entry_list, names)))
+        out <- data.frame(matrix(nrow = length(entry_list), ncol = length(names)))
+        names(out) <- names
+
+        for (i in seq_along(entry_list)) {
+            for (j in  names(entry_list[[i]])) {
+                out[i, j] <- entry_list[[i]][1, j]
+            }
+        }
+        #################
+
+        out <- as.data.frame(out, rownames = rownames, ...)
+
+        #When a bad key argument is passed to the the shoji index it returns a dataframe
+        # with an NA value. This exclude those columns.
+        exclude_cols <- grepl("^NA", names(out)) &
+            vapply(out, function(x)all(is.na(x)), FUN.VALUE = logical(1))
+        out <- out[, !exclude_cols, drop = FALSE]
+        #reorder columns to match the order in which keys were supplied
+        if (keys == TRUE) {
+            ordered_names <- names(out)
+        } else {
+            ordered_names <- names(out)[names(out) %in% keys]
+        }
+        out <- out[, ordered_names, drop = FALSE]
+
+        missing_keys <- dQuote(keys[!(keys %in% names(out))])
+
+        if (any(!(keys %in% names(out))) && keys != TRUE) {
+            if (length(missing_keys) == 1) {
+                error_text <-  " is an invalid key for catalogs of class "
+            } else {
+                error_text <- " are invalid keys for catalogs of class "
+            }
+            halt(paste0(
+                paste(missing_keys, collapse = ", ")),
+                error_text,
+                class(x),
+                ".")
+        }
+
+        return(out)
     }
-    return(out)
 }
+
+#' entry_to_df
+#' Turns an entry in the catalog into
+#'
+#' @param entry
+#' A single entry in a catalog
+#'
+#' @param list_col_names
+#' Some entries in a catalog do not fit neatly into a traditional dataframe and are
+#' instead stored in a list-column. This is controlled by passing the attribute names
+#' as the `list_col_names` argument. `
+#'
+#' @return A one row data frame
+#'
+entry_to_df <- function(entry, list_col_names ){
+            entry[vapply(entry, is.null, logical(1))] <- NA
+            vect_col <- entry[!(names(entry) %in% list_col_names)]
+            entry_df <- as.data.frame(vect_col, stringsAsFactors = FALSE)
+
+            if (any(names(entry) %in% list_col_names)) {
+                list_col <- entry[list_col_names]
+                list_df <- data.frame(matrix(nrow = 1, ncol = length(list_col)))
+                names(list_df) <- names(list_col)
+                for (i  in seq_along(list_col)) {
+                    list_df[[1, i]] <- list_col[i]
+                }
+                entry_df <- cbind(entry_df, list_df)
+            }
+            entry_df
+        }
```

Comment:
It won't fail exactly, but it will produce odd content. Here are the case: 
1) Non-list variable passed as a list variable
-  The non-list variable is returned as a list column with a single entry in each list
2) Nonsense passed to list variable argument
- No change in function behaviour
3) List variable for example `subvariable` not passed to list variables
- In this case each entry of the list will get its own column in the output database. `catalogToDataframe` will still work but just return a wider than usual dataframe. 

In terms of re-writing to pick up the list automatically. There was some discussion with Neal about this on this PR way back when. Here are my thoughts: 

1) I don't think there's a way to tell the difference between entries which are lists of length 1 and entries which are variables stored in lists of length 1. For instance if you have a variable which has only 1 sub variable, the subvariable attribute will look just like any other attribute which we don't want stored in a list column. It seems like the only way to tell those two cases apart is by hard coding the name. 
2) I think it's better for us to decide which variables _should_ be stored as list-columns in the output dataset, and either set those as the default for `catalogToDataframe()` or write `as.data.frame()` methods for other catalogs which specify the proper list-columns. To me that seems more stable than having it autodetect based on the particular catalog. Especially when you imagine someone using a post-processing script on the resulting dataframe. It's better for the dataframe variable to always have the same class than to have it change based on the catalog.